### ISSUE TITLE
Correct Color Palette

### ISSUE THREAD

**Issue by James-F2** [ekeeke/Genesis-Plus-GX#345]
Please add the correct 4bit PPU color palette to this excellent emulator.
As I understand the Genesis has 3bit and an emphasis bit, and currently the image looks darker than what the real mega drive PPU outputs.

Here is the correct LUT in the MiSTer Genesis core:
https://github.com/MiSTer-devel/Genesis_MiSTer/commit/a297b522d99adff599db91501701f9efe137a860

**Comment by ekeeke** [comment]
Well, I have compared the output of this emulator on Wii with my MD1/MD2 output on same TV, both with RGB scart and it looked quite exactly the same, and certainly not darker.

I know about the correct RGB ramp as it was first discussed on spritesmind forums long time ago (see http://gendev.spritesmind.net/forum/viewtopic.php?f=22&t=2188) but it has nothing to do with making image darker and it didn't make any noticeable differences in any games when I tried to implement it so I ditched it.

As for the 'darker' colors thing, it's more related to the fact that most recent emulators (including FPGA implementations) assume the VDP max level corresponds to 255 in RGB888 encoding scheme, i e that white output by MD console is 'pure white' as defined by video standards, while this emulator rather use max measured level ratio compared to standard video RGB level (4.7V for max level at VDP output, scaled down to 0.9Vpp at video encoder input vs 5V for VDP reference voltage which would give 1Vpp at encoder input and correspond to 'pure white' voltage in encoder datasheet). As said above, my experiments showed that (at least with my french RGB MD models) my assumptions were correct and white level output by MD was not 'pure white'. Using max RGB888 range in emulator showed very minimal difference (making screen indeed slightly brighter) but still enough difference with MD console output used as reference to make me think current model I use is the correct one.

That said, I understand this is mostly an user preference thing at the end so I am not against adding an option to let user between 'limited' and 'full' RGB range. Same for letting users choosing between 'linear' or 'original' RGB levels and play the 'spot the differences' game by themselves ;-)

**Comment by James-F2** [comment]
It made very noticeable difference when it was implemented in the MiSTer, and looks much closer to what I see from my console with the OSSC.
That would be very awesome if it will be switchable and accurate to the original in GPGX too.
Thank you ekeeke.

**Comment by ekeeke** [comment]
Could you provide info on which games it made noticeable differences to you or better, screen shots made with core before and after correct levels implementation ?

It really wasn't much noticeable from what I remember when I tried it so maybe Mister initial implementation was wrong: it is indeed the case with some older emulators which get shadow/highlight color levels wrong or use much reduced rgb range, i.e they simply shift color RAM RGB 3-bit value left to get RGB 8-bit value, thus setting max level to 224 instead of 240 in this emulator (or 255 with ful range) and indeed getting a darker image.

**Comment by James-F2** [comment]
It made difference in all games.
Mister uses very close Shadow and HiLite values from the hardware measurements you linked.
Though Kitrinx can't remember what emulator she took the values from.

Mister Aladdin Old
![Mister Aladdin Old](https://user-images.githubusercontent.com/31185814/99883329-72709d00-2c2f-11eb-8f11-411ac697cebe.png)

Mister Aladdin Correct
![Mister Aladdin Correct](https://user-images.githubusercontent.com/31185814/99883340-7a304180-2c2f-11eb-9b44-e9282b52688a.png)

Mister Sonic Old
![Mister Sonic Old](https://user-images.githubusercontent.com/31185814/99883349-887e5d80-2c2f-11eb-9664-845ef46b00e8.png)

Mister Sonic Correct
![Mister Sonic Correct](https://user-images.githubusercontent.com/31185814/99883353-8e743e80-2c2f-11eb-8044-c692d94ee9fd.png)

**Comment by James-F2** [comment]
Got it.
Blast Em.

static uint8_t levels[] = {0, 27, 49, 71, 87, 103, 119, 130, 146, 157, 174, 190, 206, 228, 255};

https://github.com/AmatCoder/blastem-gtk/blob/37cf19bfbe8963490a7919f120ab1523abc727a4/vdp.c#L67

**Comment by ekeeke** [comment]
Yes, those are the numbers derived from the voltage measurements in spritesmind thread, scaled to maximal rgb 8-bit range (255). Mister use a 16-entries lut because it's likely easier for implementation, although there are only 15 levels  in reality, hence the duplicated 255 value.

Looking at this Mister commit you linked above and Mister video mixer implementation, it seems that Mister older implementation was using same max value as Genesis Plus GX (238 i.e 11101110b) and I can confirm the differences you see in those pictures are similar to the ones I noticed on my TV when trying out 255 as RGB max value (instead of 238) and I remember testing the exact same games. But like I said also, comparison with my RGB MD models on the same TV showed that 238 as max RGB value was more correct and gave closer results to my original console. Maybe it's different when using non-RGB models which output composite video encoded from VDP RGB output ?

**Comment by James-F2** [comment]
Can't say about Composite because I haven't used that in ages.
But I can say that the Mister has Analog RGB output that is connected to my european 29" Sony Trinitron with a VGA to Scart cable, and the current Mister palette is very similar to the Genesis when I switch between the two with my Scart switcher.
The real Genesis might be slightly brighter, but the resulting gamma curve is very similar,, certainly closer than the old mister implementation.

Anyway, it would be very useful to have the original measured levels available in GPGX.

**Comment by ekeeke** [comment]
>Can't say about Composite because I haven't used that in ages.

What I meant is that unless you are using a French RGB Megadrive with official RGB scart, your console still outputs a composite video signal that is converted back to RGB by an external hardware converter. Depending how these recreated RGB signals are amplified by those converters (and they likely are amplified to match RGB scart spec), the end result could be brighter than with original RGB consoles (which uses its own official RGB adapter box).

Even RGB-modded consoles, which use third-party circuitry to bypass the video encoder and adapt the VDP RGB output for RGB scart standard probably have different amplification levels than the official RGB console.

That said, I understand that people like the slightly brighter image those modern RGB solutions provide and also, most emulator users run them through modern screens, where a given RGB range maybe looks different than with CRT TVs,  so the most logical is indeed to provide both options (limited or full RGB) in the emulator.

**Comment by James-F2** [comment]
Thank you for considering to implement this option.

I think it has to do more with 'linear' vs 'original' grey scale, than 'limited' vs 'full' range.
I don't mind slightly darker whites, but I do mind if the emulator does not match the original grey scale.
Maybe I'm misunderstanding this, and they are one and the same.

**Comment by p1pkin** [comment]
@James-F2 
> but I do mind if the emulator does not match the original VDP grey scale tracking.

we don't know how exactly original VDP works,  there required chip decap and its analysis to get all the details.
so, at this point it is more like "one approximation VS another approximation", where one is not any "better" or "more accurate" than another.

**Comment by James-F2** [comment]
The equivalent to 'Full Range' + 'Original Grey' most certainly looks closer to what the Genesis outputs than what GPGX has now.
It is also how the Genesis looks with Scart RGB + OSSC.
BlastEm and MiSTer use that approach too, because it looks authentic to the real console output.

I hope to see that implemented in GPGX soon.

**Comment by eadmaster** [comment]
The Master System also has [a non-linear blue channel](https://www.retrorgb.com/sega-master-system-non-linear-blue-channel-findings.html) that should be reproduced.

**Comment by mikepavone** [comment]
> BlastEm and MiSTer use that approach too, because it looks authentic to the real console output.

For what it's worth, I did not pick 255 as the max level in BlastEm because I did any real comparison to hardware. My general thinking here is that since max brightness can vary significantly between displays (especially since most displays have brightness controls) that there is not really any ground truth when it comes to max brightness and so I went with 255 because it was my personal preference.

There is some more objective truth with respect to the relative values, but as has already been pointed out the non-linearity is relatively subtle compared to the choice of max brightness. Further, what I'm doing in BlastEm is not really fully correct either as NTSC and PAL have different gamma curves than an sRGB display and I do not currently account for that at all.

**Comment by birdybro** [comment]
I agree with @ekeeke that a French Megadrive doing SCART RGB is going to differ significantly from one that uses an RGB modification on it. Additionally even different regions had different levels, see this article for more information @James-F2 :

https://medium.com/rgb-inside/comparing-native-rgb-video-quality-of-different-models-of-mega-drive-and-sega-genesis-english-e3d3703270ba

I have a JVC X'Eye, MegaJet, an MD1VA3 and an MD1VA6. I also grew up with a 32x on top of a Genesis Model 2. I can confirm they all look different. So which one is "correct"? The answer is, none. Here's the variety of RGB output compared in the article so you don't have to scroll through it and so it doesn't disappear when the article goes away someday:

Japanese MD1
![jpmd1](https://user-images.githubusercontent.com/16388068/160263640-d3d726c5-c237-4acc-bc3f-358ceab80e34.png)

Japanese MD2
![jpmd2](https://user-images.githubusercontent.com/16388068/160263641-42cabe7c-fcc4-4b55-a018-f9f9350aa939.png)

US MD1 "High Definition"
![usmd1](https://user-images.githubusercontent.com/16388068/160263649-d8d8defd-e0c8-4877-a7fc-052e70e74c82.png)

US MD1 no "High Definition" written on it
![usmd1_2](https://user-images.githubusercontent.com/16388068/160263656-3df3107a-7ee2-4b33-b94a-1814913d269f.png)

US MD1 Modded by cutting the leg of the composite video signal to clean up RGB
![usmd1_modded](https://user-images.githubusercontent.com/16388068/160263674-86a84d76-d779-4372-bf27-777032582826.png)

Tec Toy Brazilian MD3 (actually an MD2 but with a Fujitsu video encoder)
![brazilmd3](https://user-images.githubusercontent.com/16388068/160263690-e9181b50-05ba-4423-bdd7-52eae7a96b61.png)

Tec Toy Brazilian CDX
![bztectoycdx](https://user-images.githubusercontent.com/16388068/160263694-08ef514a-6ee1-4892-af84-9f1482d45d9a.png)

32x (yes it modifies the image quality)
![32x](https://user-images.githubusercontent.com/16388068/160263697-b21274b7-b728-404d-b63d-6972f4f13127.png)

Mega Jet
![megajet](https://user-images.githubusercontent.com/16388068/160263705-c5d10c19-04de-42c4-8a5b-36ccdf0d565e.png)

JVC X'Eye
![jvcxeye](https://user-images.githubusercontent.com/16388068/160263706-68b4b9b3-ed3b-44fe-a886-90115804ef46.png)

The likely solution would be to have a variety of color palettes to be selectable like other emulators have done before (usually NES emulators, because of that system's uniquely strange color situation). You can really get into the weeds on this stuff. The audio differences between every variation of Sega Genesis are a ridiculously long list. :P

**Comment by Dogway** [comment]
To add to the conversation I don't know how accurate BlastEm is regarding color output but from some tests it's nothing fancy, just lowering contrast in a sigmoidal fashion in RGB, so saturation is decreased along luma contrast.

pseudo-call:
`sigmoidal_inv(rgb, strength=4.0, pivot=0.52)`

I added an option to mimic the look in [grade](https://github.com/libretro/slang-shaders/blob/master/misc/shaders/grade.slang) shader, along the blue push for SMS. Will upload in a few days.

**Comment by mikepavone** [comment]
Literally all I did in BlastEm was took the voltages others had measured for each of the Mode 5 RGB levels, divided by the max measured voltage and multiplied by 255. Those values are in `levels` in vdp.c and get used for translating palette entries. Nothing fancy at all.

**Comment by Dogway** [comment]
Well so it seems legit, at least it's based on measurements. Would be nice to add source reference so we (ekeeke included) can have a look and maybe implement it.

**Comment by ekeeke** [comment]
@Dogway: source is already referenced, see spritesmind forum link I posted in second message. The values used in Blastem are also referenced somewhere above. As Mask of Destiny explained, there is nothing complicated, it's just that I never took the time implementing it (it also requires some design changes in vdp rendering code). This will be done someday, eventually...

**Comment by KrokusPokus** [comment]
@ekeeke: Is adding an option for a 255 max level still on the table? My main concern is using the rather deplorable backlight output of my laptop's TFT screen to its fullest. With a max of 238, a good 7% of the light output gets wasted. (Also, with back-lit displays in general, having to turn up the backlight to compensate for not using full range increases backlight bleeding, making blacks more greyish...)

**Comment by Jobima1st** [comment]
seems the best thing to do is what @birdybro said "have a variety of color palettes to be selectable like other emulators have done before (usually NES emulators, because of that system's uniquely strange color situation)"

**Comment by Clownacy** [comment]
There is such a thing as ['too many options'](https://ometer.com/preferences); sometimes, it is best for software to just stick with a sane default.
